File: Day065_2025/Q065_GFG.java
class Q065_GFG {
    public static int lcs(String s1, String s2) {
        int n = s1.length();
        int m = s2.length();
        
        int[] next = new int[m + 1];
        int[] cur = new int[m + 1];

        for (int i = n - 1; i >= 0; i--) {
            for (int j = m - 1; j >= 0; j--) {
                if (s1.charAt(i) == s2.charAt(j))
                    cur[j] = 1 + next[j + 1];
                else
                    cur[j] = Math.max(next[j], cur[j + 1]);
            }
            next = cur.clone();
        }

        return next[0];
    }
}

File: Day065_2025/Q065_GFG.py
class Solution:
    def lcs(self, s1: str, s2: str) -> int:
        n, m = len(s1), len(s2)
        
        next_row = [0] * (m + 1)
        cur = [0] * (m + 1)

        for i in range(n - 1, -1, -1):
            for j in range(m - 1, -1, -1):
                if s1[i] == s2[j]:
                    cur[j] = 1 + next_row[j + 1]
                else:
                    cur[j] = max(next_row[j], cur[j + 1])
            next_row = cur[:]

        return next_row[0]
